### PR TITLE
[UIF-429] - Introducing error code for scenario

### PR DESCRIPTION
**Pull request by azizbekxm**
## Purpose

https://issues.folio.org/browse/UIF-429
## Approach

Introduced budgetNotFoundForTransaction error Code to avoid genericError
## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all the appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?
- Did you modify code to call some additional endpoints?
  - [ ] If so, do you check that necessary module permission added in ModuleDescriptor-template.yaml file?

Ideally, all the PRs involved in breaking changes would be merged on the same day
to avoid breaking the folio-testing environment.
Communication is paramount if that is to be achieved,
especially as the number of inter-module and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems,
ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/service/budget/BudgetService.java [hunk 1]
```diff
@@ -147,8 +145,9 @@ public Future<Budget> getBudgetByFundIdAndFiscalYearId(String fiscalYearId, Stri
           handleFailure(promise, reply);
         } else if (reply.result()
           .isEmpty()) {
-          logger.error(BUDGET_NOT_FOUND_FOR_TRANSACTION);
-          promise.fail(new HttpException(Response.Status.BAD_REQUEST.getStatusCode(), BUDGET_NOT_FOUND_FOR_TRANSACTION));
+          logger.error(BUDGET_NOT_FOUND_FOR_TRANSACTION.toError());
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/service/budget/BudgetService.java:148]: Can you add to logs fiscal year id and fund Id, you can use in logs your custom message

Comment (azizbekxm) [reply src/main/java/org/folio/service/budget/BudgetService.java:148]: Ok, I will fixt

File: src/main/java/org/folio/service/budget/BudgetService.java [hunk 2]
```diff
@@ -147,8 +145,9 @@ public Future<Budget> getBudgetByFundIdAndFiscalYearId(String fiscalYearId, Stri
...
+          promise.fail(new HttpException(Response.Status.BAD_REQUEST.getStatusCode(),
```
[the remark this answers is not in the document]
Comment (azizbekxm) [reply src/main/java/org/folio/service/budget/BudgetService.java:149]: Ok, I will use

File: src/main/java/org/folio/service/budget/BudgetService.java [hunk 3]
```diff
@@ -147,8 +145,9 @@ public Future<Budget> getBudgetByFundIdAndFiscalYearId(String fiscalYearId, Stri
           handleFailure(promise, reply);
         } else if (reply.result()
           .isEmpty()) {
-          logger.error(BUDGET_NOT_FOUND_FOR_TRANSACTION);
-          promise.fail(new HttpException(Response.Status.BAD_REQUEST.getStatusCode(), BUDGET_NOT_FOUND_FOR_TRANSACTION));
+          logger.error("Budget not found for fundId: {} and fiscalYearId: {}", fundId, fiscalYearId);
+          promise.fail(new HttpException(Response.Status.NOT_FOUND.getStatusCode(),
+            JsonObject.mapFrom(BUDGET_NOT_FOUND_FOR_TRANSACTION.toError()).encodePrettily()));
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/service/budget/BudgetService.java:150]: BUDGET_NOT_FOUND_FOR_TRANSACTION - it is just string message, can we omit using JsonObject?

Comment (azizbekxm) [reply src/main/java/org/folio/service/budget/BudgetService.java:150]: it actually not string anymore, in new case I introduce error code mod-finance-storage also, it is now giving Error object we should use JsonObject I guess
image

Comment (SerhiiNosko) [reply src/main/java/org/folio/service/budget/BudgetService.java:150]: I seem this approach also used in all other places, so I am ok with that

**Review by SerhiiNosko — APPROVED** [review]

**Review by Abdulkhakimov — APPROVED** [review]

**Pull request merged**